jobrunner/job_executor.py

```diff
@@ -33,26 +45,86 @@ class JobDefinition:
     allow_database_access: bool  # whether this job should have access to the database
 
 
+class ExecutorState(Enum):
+    PREPARING = "preparing"
+    PREPARED = "prepared"
+    EXECUTING = "executing"
+    EXECUTED = "executed"
+    FINALIZING = "finalizing"
+    FINALIZED = "finalized"
+    GONE = "gone"
+    ERROR = "error"
+
+
+@dataclass
+class JobStatus:
+    state: ExecutorState
+    message: Optional(str)
+
+
 @dataclass
 class JobResults:
-    state: State
-    status_code: Optional[StatusCode]
-    status_message: str
     outputs: Mapping[str, str]
+    unmatched: List[str]
     exit_code: int
     image_id: str
 
 
 class JobAPI:
-    def run(self, job: JobDefinition) -> None:
+    """
+    API for managing job execution.
+
+    This API is called by the job-runner to manage each job it is tracking. It models the running of a job as a state
+    machine,and the methods below are the transitions between states.
+
+    Given the long running nature of jobs, it is an asychronous API, and calls should not block for a more than a few
+    seconds, returning a transition to the next state.
+
+    All the state transition methods (prepare(), execute(), finalize(), terminate(), cleanup()) should be idempotent. If
+    the relevent task they are responsible for is already running for that job, they should not start a new task, and
+    instead return successfully the current state.
```

Comment:
Also I think we need a section explaining that any async errors (but not failure of the job itself) should be reacted to by moving to `ERROR` state (and returning a helpful message when the status is next retrieved).

Reply:
Good call, will add